feat(agw): Adding Ansible role and playbook to install 5G AGW

tapasmishra opened the pull request:
## Summary

Ansible script to install 5G AGW on bare metal or VM with minimal information. 

## Test Plan

Included in the README file

Here is the installation log file: [AGW Installation log](https://github.com/wavelabsai/agw_install_log/blob/master/ansible_agw_log.txt)

## Additional Information

This ansible playbook will help user's to install and configure a 5G AGW. Also user's can test any version of the build Artifact from the test repo.

tapasmishra committed:
Adding ansible role and playbook to install 5G AGW

tapasmishra committed:
Adding OS version check and hardware specifications

pshelar reviewed (changes requested):
Can you modify existing AGW installation script with 5G option to enable 5G on the deployment. I dont see need to have different deployment for 4G and 5G.

tapasmishra committed:
Adding instructions for 5G and 4G version installations

tapasmishra commented:
> Can you modify existing AGW installation script with 5G option to enable 5G on the deployment. I dont see need to have different deployment for 4G and 5G.

This Ansible plyabook is not just to deploy 5G. If user need then he/ she can install 4G as well by modifying few attributes. I have updated the REAME with instructions. The existing AGW installation script includes a shell along with Ansible. This new approach will eliminate the use of additional shell script. The reason for to place in a different location is to allow user's to test it first. If everything will look good then I will update the PR to replace existing approach along with changes to the Documents.

tapasmishra committed:
Updated REAME with informations

milapjoshi commented on lte/gateway/deploy/magma_5g_agw/ansible_vars.yaml:
```diff
@@ -0,0 +1,19 @@
+# eth1_gateway: ""
+
+magma_control_proxy_path: "/home/ubuntu/agw-deployment/ansible/roles/magma-agw/files/control_proxy.yaml"
+magma_rootCA_path: "/home/ubuntu/agw-deployment/ansible/roles/magma-agw/files/rootCA.pem"
```
Which orc8r this rootCA pointing to. can we not hardcode rootCA or control_proxy

tapasmishra replied:
Thanks, it was by mistake. Removed the hardcoded path.

milapjoshi commented on lte/gateway/deploy/magma_5g_agw/roles/magma-agw/tasks/config5g.yaml:
```diff
@@ -0,0 +1,62 @@
+  copy:
+    src: roles/magma-agw/files/mme.yml
+    dest: "/var/opt/magma/configs/mme.yml"
+  register: mme_yml_file
```
can you use process config files present at /etc/magma/ and `sed` required changes.

tapasmishra replied:
Yes, we can process but as per the developer's it's always good to overried the values by putting the config file inside `/var/opt/magma/configs`. It will give you two benefits: 1. Configuration's will be remain same even after a package upgrade. 2. By removing the configurations you can disable the 5G mode. If you still want to edit the configuration of /etc/magma then I will work on it.

milapjoshi commented on lte/gateway/deploy/magma_5g_agw/roles/magma-agw/tasks/install-magma.yaml:
```diff
@@ -0,0 +1,176 @@
+- name: Start service openvswitch-switch.
+  become: true
+  service:
+    name: openvswitch-switch
```
we need to make sure ovs has the required kernel pkg installed. there is this script ` ovs-kmod-upgrade.sh ` as a part of upgrade procedure. we need to make sure we use that upgrade script

tapasmishra replied:
Added logic in the Ansible role to take care of the issue.

milapjoshi reviewed (changes requested):
please see some comments.

tapasmishra committed:
Adding logic to validate kernel version

pshelar commented on lte/gateway/deploy/magma_5g_agw/README.md:
````diff
@@ -0,0 +1,111 @@
+eth1_netmask: ""
+eth0_gateway: ""
+eth1_gateway: ""
+```
````
I dont think it is right way to configure ansible variables, you can use `ansible-playbook -e` option to pass these variables.

tapasmishra replied:
With -e option either you can either add variables with key=value or YAML/JSON, if filename prepend with @. 

https://docs.ansible.com/ansible/latest/cli/ansible-playbook.html#cmdoption-ansible-playbook-e

pshelar commented:
> > Can you modify existing AGW installation script with 5G option to enable 5G on the deployment. I dont see need to have different deployment for 4G and 5G.
> 
> This Ansible plyabook is not just to deploy 5G. If user need then he/ she can install 4G as well by modifying few attributes. I have updated the REAME with instructions. The existing AGW installation script includes a shell along with Ansible. This new approach will eliminate the use of additional shell script. The reason for to place in a different location is to allow user's to test it first. If everything will look good then I will update the PR to replace existing approach along with changes to the Documents.



> > Can you modify existing AGW installation script with 5G option to enable 5G on the deployment. I dont see need to have different deployment for 4G and 5G.
> 
> This Ansible plyabook is not just to deploy 5G. If user need then he/ she can install 4G as well by modifying few attributes. I have updated the REAME with instructions. The existing AGW installation script includes a shell along with Ansible. This new approach will eliminate the use of additional shell script. The reason for to place in a different location is to allow user's to test it first. If everything will look good then I will update the PR to replace existing approach along with changes to the Documents.

magma AGW is suppose to be converged core. so the deployment for 4G and 5G should be exactly same.
User should be able to enable or disable 5G using agw config files.

pshelar commented on lte/gateway/deploy/magma_5g_agw/roles/magma-agw/files/gateway_mconfig:
```diff
@@ -0,0 +1,157 @@
+{
```
if you really want to go down this path of separate deployment role, lets eliminate duplicate config files.
The mconfig can be update at time of deployment.

tapasmishra replied:
Ahh .. sorry its not required at all. I was doing some research and left this file unintentionally here in this PR. Let me remove it.

tapasmishra replied:
Updated.

tapasmishra commented:
> 

This is what exactly this cookbook is doing. It's adding couple of config files while deploying the latest artifacts to enable 5G if the user want's to enable 5G. There is no difference in the installation of the AGW for 4G or 5G.

tapasmishra committed:
Removing unnecessary file inside files directory'

tapasmishra committed:
Merge branch 'magma:master' into master

quentinDERORY reviewed (changes requested):
Would be better to have an option/sets of config in the 4g deployment to enable 5g instead of  creating whole new playbook/role

tapasmishra commented:
> Would be better to have an option/sets of config in the 4g deployment to enable 5g instead of creating whole new playbook/role

@quentinDERORY The purpose of bringing a new role/ playbook is not just to install 4G and 5G of AGW. Also, to reduce the complexity of the installation. This will help to reduce the prerequisites that we have defined today. User's needs to just bring hardware as specified in the README with two ethernet ports having Ubuntu 20.04. No more customisation is required during the OS installation. Additionally, the cookbook will help users to configure the AGW for a given orc8r if they want.

pshelar commented:
> > Would be better to have an option/sets of config in the 4g deployment to enable 5g instead of creating whole new playbook/role
> 
> @quentinDERORY The purpose of bringing a new role/ playbook is not just to install 4G and 5G of AGW. Also, to reduce the complexity of the installation. This will help to reduce the prerequisites that we have defined today. User's needs to just bring hardware as specified in the README with two ethernet ports having Ubuntu 20.04. No more customisation is required during the OS installation. Additionally, the cookbook will help users to configure the AGW for a given orc8r if they want.

what customization is required for 4G? can you explain what exact improvements are?

tapasmishra commented:
> > > Would be better to have an option/sets of config in the 4g deployment to enable 5g instead of creating whole new playbook/role
> > 
> > 
> > @quentinDERORY The purpose of bringing a new role/ playbook is not just to install 4G and 5G of AGW. Also, to reduce the complexity of the installation. This will help to reduce the prerequisites that we have defined today. User's needs to just bring hardware as specified in the README with two ethernet ports having Ubuntu 20.04. No more customisation is required during the OS installation. Additionally, the cookbook will help users to configure the AGW for a given orc8r if they want.
> 
> what customization is required for 4G? can you explain what exact improvements are?

@pshelar I have documented in the Readme file about how to install 4G using the playbook. Let me know if those informations are not sufficient then I can work on it to improve. Below are the improvements I am trying to bring:

- Removing the complexity of shell script along with Ansible
- Removing the complexity of having different scripts for installation and up-gradation
- Removing manual effort of AGW configuration
- Added functionality to install and test CI builds
- Added functionality to downgrade the AGW version 

Requesting all of you to try the script once to install/ upgrade your AGW and then if any enhancement is required then we can do it.

shanku9 commented:
@tapasmishra 
- Can you add steps to install Ansible on the end host? Earlier this step was wrapped in the bash script and was transparent to end user
- SSH Key based auth is vital for script execution. Please elaborate steps to enable this process on local as well as remote host   
fatal: [10.253.1.49]: FAILED! => {
    "msg": "Timeout (602s) waiting for privilege escalation prompt: "
}
https://github.com/magma/magma/pull/11319/files#diff-1c4308fedbbea2b9236332356d757cdfb85c35a0df506b3982ef1424ae295b49R41
- Add Troubleshooting Guide

tapasmishra commented:
> @tapasmishra
> 
> * Can you add steps to install Ansible on the end host? Earlier this step was wrapped in the bash script and was transparent to end user
> * SSH Key based auth is vital for script execution. Please elaborate steps to enable this process on local as well as remote host
>   fatal: [10.253.1.49]: FAILED! => {
>   "msg": "Timeout (602s) waiting for privilege escalation prompt: "
>   }
>   https://github.com/magma/magma/pull/11319/files#diff-1c4308fedbbea2b9236332356d757cdfb85c35a0df506b3982ef1424ae295b49R41
> * Add Troubleshooting Guide

@shanku9  This is really a good idea. Let me work on it will update the PR soon with the changes.

tapasmishra committed:
Updating README as per the PR comments.

tapasmishra committed:
Merge branch 'master' of https://github.com/wavelabsai/magma

tapasmishra committed:
Updating Readme with few fixes..

tapasmishra commented:
> > @tapasmishra
> > 
> > * Can you add steps to install Ansible on the end host? Earlier this step was wrapped in the bash script and was transparent to end user
> > * SSH Key based auth is vital for script execution. Please elaborate steps to enable this process on local as well as remote host
> >   fatal: [10.253.1.49]: FAILED! => {
> >   "msg": "Timeout (602s) waiting for privilege escalation prompt: "
> >   }
> >   https://github.com/magma/magma/pull/11319/files#diff-1c4308fedbbea2b9236332356d757cdfb85c35a0df506b3982ef1424ae295b49R41
> > * Add Troubleshooting Guide
> 
> @shanku9 This is really a good idea. Let me work on it will update the PR soon with the changes.
@shanku9 I have updated the README file as per your suggestions. I am working on the Troubleshooting guid and I will update the repository very soon.

shanku9 reviewed (commented):
@tapasmishra 
Install not working 
`fatal: [10.253.1.90]: FAILED! => {"msg": "Timeout (62s) waiting for privilege escalation prompt: "}`
[Logs](https://gist.github.com/shanku9/2dc0b08699804d5b06c9a138bd9ef974) 

This install process requires user to update several files **manually** before actual AGW installation compared to existing 1 click [deployment](https://github.com/magma/magma/blob/master/lte/gateway/deploy/agw_install_ubuntu.sh) 
Also there are many overlaps with existing process
[agw_pre_check_ubuntu.sh](https://github.com/magma/magma/blob/master/lte/gateway/deploy/agw_pre_check_ubuntu.sh)
[agw_post_install_ubuntu.sh](https://github.com/magma/magma/blob/master/lte/gateway/deploy/agw_post_install_ubuntu.sh)

Can we update existing [install process](https://github.com/magma/magma/blob/master/lte/gateway/deploy/agw_install_ubuntu.sh#L209) and enable [5G specific options](https://gist.github.com/shanku9/60a37bf6ce892f9cd90a093009af1f3b)

pshelar commented:
> > > > Would be better to have an option/sets of config in the 4g deployment to enable 5g instead of creating whole new playbook/role
> > > 
> > > 
> > > @quentinDERORY The purpose of bringing a new role/ playbook is not just to install 4G and 5G of AGW. Also, to reduce the complexity of the installation. This will help to reduce the prerequisites that we have defined today. User's needs to just bring hardware as specified in the README with two ethernet ports having Ubuntu 20.04. No more customisation is required during the OS installation. Additionally, the cookbook will help users to configure the AGW for a given orc8r if they want.
> > 
> > 
> > what customization is required for 4G? can you explain what exact improvements are?
> 
> @pshelar I have documented in the Readme file about how to install 4G using the playbook. Let me know if those informations are not sufficient then I can work on it to improve. Below are the improvements I am trying to bring:
> 
> * Removing the complexity of shell script along with Ansible
> * Removing the complexity of having different scripts for installation and up-gradation
> * Removing manual effort of AGW configuration
> * Added functionality to install and test CI builds
> * Added functionality to downgrade the AGW version
> 
> Requesting all of you to try the script once to install/ upgrade your AGW and then if any enhancement is required then we can do it.



> > > > Would be better to have an option/sets of config in the 4g deployment to enable 5g instead of creating whole new playbook/role
> > > 
> > > 
> > > @quentinDERORY The purpose of bringing a new role/ playbook is not just to install 4G and 5G of AGW. Also, to reduce the complexity of the installation. This will help to reduce the prerequisites that we have defined today. User's needs to just bring hardware as specified in the README with two ethernet ports having Ubuntu 20.04. No more customisation is required during the OS installation. Additionally, the cookbook will help users to configure the AGW for a given orc8r if they want.
> > 
> > 
> > what customization is required for 4G? can you explain what exact improvements are?
> 
> @pshelar I have documented in the Readme file about how to install 4G using the playbook. Let me know if those informations are not sufficient then I can work on it to improve. Below are the improvements I am trying to bring:
> 
> * Removing the complexity of shell script along with Ansible
> * Removing the complexity of having different scripts for installation and up-gradation
> * Removing manual effort of AGW configuration
> * Added functionality to install and test CI builds
> * Added functionality to downgrade the AGW version
> 
Better way would be to add incremental changes (using separate PRs) for 4G installation for each of this point, once that is done, use same method for 5G installation.


> Requesting all of you to try the script once to install/ upgrade your AGW and then if any enhancement is required then we can do it.

pshelar reviewed (changes requested):
as mentioned above, create separate PRs for each improvement for existing installation before adding 5G installation.
